library/src/armory/metrics/detection.py

```diff
@@ -0,0 +1,189 @@
+            num_gt_boxes = len(gt_boxes)
+
+            # Initialize arrays that will indicate whether each respective ground-truth
+            # box is a true positive or misclassified
+            true_positive_array = torch.zeros((num_gt_boxes,))
+            misclassification_array = torch.zeros((num_gt_boxes,))
+
+            # Only consider the model's confident predictions
+            assert y_pred["scores"] is not None
+            conf_pred_indices = torch.where(y_pred["scores"] > self.score_threshold)[0]
+            if self.class_list:
+                # Filter out predictions from classes not in class_list kwarg
+                conf_pred_indices = conf_pred_indices[
+                    torch.isin(y_pred["labels"][conf_pred_indices], self.class_list)
+                ]
+
+            # For each confident prediction
+            for y_pred_idx in conf_pred_indices:
+                y_pred_box = y_pred["boxes"][y_pred_idx]
+
+                # Compute the iou between the predicted box and the ground-truth boxes
+                ious = (
+                    intersection_over_union(
+                        torch.stack([y_pred_box]), gt_boxes, aggregate=False
+                    )[0]
+                    if len(gt_boxes) > 0
+                    else torch.tensor([])
+                )
+
+                # Determine which ground-truth boxes, if any, the predicted box overlaps with
+                overlap_indices = torch.where(ious > self.iou_threshold)[0]
+
+                # If the predicted box doesn't overlap with any ground-truth boxes, increment
+                # the hallucination counter and move on to the next predicted box
+                if len(overlap_indices) == 0:
+                    num_hallucinations += 1
+                    continue
+
+                # For each ground-truth box that the prediction overlaps with
+                for y_idx in overlap_indices:
+                    # If the predicted label is correct, mark that the ground-truth
+                    # box has a true positive prediction
+                    if gt_labels[y_idx] == y_pred["labels"][y_pred_idx]:
+                        true_positive_array[y_idx] = 1
+                    else:
+                        # Otherwise mark that the ground-truth box has a misclassification
+                        misclassification_array[y_idx] = 1
+
+            true_positive_rate = (
+                true_positive_array.mean()
+                if len(true_positive_array) > 0
+                else torch.tensor(0)
+            )
+            misclassification_rate = (
+                misclassification_array.mean()
+                if len(misclassification_array) > 0
+                else torch.tensor(0)
+            )
+
+            # Any ground-truth box that had no overlapping predicted box is considered a
+            # disappearance
+            disappearance_rate = 1 - true_positive_rate - misclassification_rate
```

Comment:
I realize this is copied from Armory, but I think we should define disappearance_rate to be 0 if there are no ground truth boxes (```len(targets)==0```).  Currently, disappearance rate is 1 if there are no targets, which doesn't make sense to me.  This would also require updating the unit test for the example with the empty target.